### DISCUSSION TITLE
Clock drift,recovery, etc...

### DISCUSSION THREAD

**Discussion by KC5VDJ** [kej715/DtCyber#61] [General]
I am thinking that the clock might actually be my starting point after two issues I have noted; first, that there is clock drift within nos 2, a significant chunk of a minute after some hours of runtime; and second, the fast-forward clock recovery upon host wake-up seems to only work once, and slow because it trips the rollouts and other scheduled jobs continuously until the clock is updated to current (or not (twice)).  Also, I have always admired the measureware hardware in pa-risc, combined with kernel instrumentation to interface to the performance counters, and any such instrumentation in dtcyber would obviously require a clock hook.

The issue with the scheduled jobs all running, etc in a graceful wake-up recovery is good in some cases, but results in a spate of disk writes continuously until current time, and despite Samsung's guarantee on EVOs, well, dtcyber doesn't factor ubiquitous flash SSDs.  It was written in a spinning disk era.  Optimizing for flash SSDs is another potential project I am considering.  Back to the reliable power problem...consider laptops.  Desktop or pi use could require a UPS or a big chunk of ram respectively, so problems to solve for universality....not all use a decade old laptop with inherently reliable power.  It'll either be a buffering or ramdisk thing, with associated problems, and longer shutdown saving only changed blocks or something.

**Comment by kej715** [comment]
Please be aware that realtime clock emulation is dependent upon the
platform on which DtCyber is running. In particular, the code for Windows
is quite different from the code for Unix/Linux. Are you running DtCyber on
Windows? Quite possibly, drift could occur on that host platform. FYI ...
the Cyber 865 at the Nostalgic Computing Center
<http://www.nostalgiccomputing.org> runs NOS 2.8.7, and DtCyber is hosted
on MacOS for that machine. NOS 2.8.7 literally runs for weeks in that
environment without being re-deadstarted, and no clock drift whatsoever
occurs.

**Reply by KC5VDJ** [reply]
That explains it, windows.

**Reply by KC5VDJ** [reply]
I still would like to do the perf monitoring hook and deal with the sleep/wakeup thing.  Just had to dtcyber shutdown rather than wait for it to catch up 12 hours in that fast forward thing.

**Comment by KC5VDJ** [comment]
Is there a way to get to something akin to acpi-regisrered high resolution timers in winblowz or arm/riscv under Linux / bsd?  I have a few pi boards but I forget about those, iirc there are several such timers available under arm, but the b3+ and zeroes that I have have to be polled, thus limiting speed unless I misread and clock ints are avail...  Such timers may be a bit more monotonous than the standard winblowz functions which have obvious drift, and much higher resolution, dependent on parts quality, age, and temperature (limitations of quartz).

Would speedstep be affecting the QueryPerformanceCounter()?  Is this an Intel counter or a winblowz thing?  Either way, if speedstep (c-states too?) are the issue, a monotonous acpi timer seems a better idea.  I run dtcyber on a haswell laptop, thermals suck already, fresh conductonaut this past weekend, I'm not turning off speedstep, but there is no excuse for bad wall clock time, so....  Delving deeper...

**Comment by kej715** [comment]
The current implementation in the rtc.c module uses Windows QPC (Query
Performance Counter) API. Perhaps the GetSystemTimePreciseAsFileTime
<https://learn.microsoft.com/en-us/windows/win32/api/sysinfoapi/nf-sysinfoapi-getsystemtimepreciseasfiletime>
API
would produce less drift and behave more like the gettimeofday() API used
in Unix/Linux.
